File: src/scanner.rs
use super::token::Token;
use super::token::LiteralValue;
use super::token_type::TokenType;
use super::keywords::KEYWORDS;

pub struct Scanner { 
    source: String,
    tokens: Vec<Token>,
    start: i32,
    current: i32,
    line: usize,
}

impl Scanner {

    pub fn new(source: &str) -> Scanner {
        
        Scanner {
            source: String::from(source),
            tokens: Vec::new(),
            start: 0,
            current: 0,
            line: 1
        }
        
    }

    pub fn scan_tokens(&mut self) -> &Vec<Token> {

        while !self.is_at_end() {
            self.start = self.current; 
            self.scan_token();
        }

        self.tokens.push(Token::new(
            TokenType::EOF,
            &String::from(""),
            None,
            self.line,
        ));

        &self.tokens
    }


    fn scan_token(&mut self) {

        let c = self.advance();
       
        let token_type = {
            match c {
                
                // Single Letter Tokens
                '(' => TokenType::LeftParen,
                ')' => TokenType::RightParen,
                '{' => TokenType::LeftBrace,
                '}' => TokenType::RightBrace,
                ',' => TokenType::Comma,
                '.' => TokenType::Dot,
                '-' => TokenType::Minus,
                '+' => TokenType::Plus,
                ';' => TokenType::Semicolon,
                '*' => TokenType::Star,
                
                // Double Letter Tokens
                '!' => self.conditional_select('=', TokenType::BangEqual, TokenType::Bang),
                '=' => self.conditional_select('=', TokenType::EqualEqual, TokenType::Equal),
                '<' => self.conditional_select('=', TokenType::LessEqual, TokenType::Less),
                '>' => self.conditional_select('=', TokenType::GreaterEqual, TokenType::Greater),
                
                // Comments
                '/' => {
                    if self.match_next_token('/') {
                        while self.peek() != '\n' && !self.is_at_end() { self.advance(); }
                        TokenType::NOP
                    } else {
                        TokenType::Slash
                    }
                }
               
                // Strings
                '"' => {
                    self.string();
                    TokenType::NOP
                },


                // Meaningles characters
                ' ' | '\r' | '\t' => TokenType::NOP,
                '\n' => {
                    self.line += 1;
                    TokenType::NOP
                },

                // Digits
                _ if c.is_digit(10) => {
                    self.number(); 
                    TokenType::NOP
                },

                // Alphanumeric Identifiers
                _ if c.is_alphabetic() => {
                    self.identifier();
                    TokenType::NOP
                },

                // Unknown Character
                _ =>  { 
                    eprintln!("[line {}] Error: {}", self.line, "Unexpected Character.");
                    std::process::exit(65)
                }
            }
        };
        
        if token_type != TokenType::NOP {
            self.add_token(token_type);
        }
    }
    
    // Gather Identifiers
    fn identifier(&mut self) {
        while self.peek().is_alphanumeric() { self.advance(); }
        let value = &self.source[self.start as usize..self.current as usize];
        let identifier_type= (*KEYWORDS).get(&value).unwrap_or(&TokenType::Identifier);
        self.add_token(*identifier_type);
    }

    // Gather Numbers
    fn number(&mut self) {
       while self.peek().is_digit(10) {
           self.advance();
       }

       // check for float
       if self.peek() == '.' && self.peek_next().is_digit(10) {
           self.advance();
           while self.peek().is_digit(10) { self.advance(); }
       }

        let value = self.source[self.start as usize..self.current as usize]
            .parse::<f64>()
            .expect("Failed to parse number");

        self.add_token_with_literal(TokenType::Number, Some(LiteralValue::Number(value)));
    }

    // Peek the input after the current
    fn peek_next(&mut self) -> char {
        if self.current as usize + 1 >= self.source.len() { return '\0'; }
        return self.source.chars().nth(self.current as usize + 1).unwrap();
    }

    // Gather String literals
    fn string(&mut self) {
        
       while self.peek() != '"' && !self.is_at_end() {
           if self.peek() == '\n' {
               self.line += 1
           }
           self.advance();
       }

        if self.is_at_end() {
            eprintln!("[line {}] Error: {}", self.line, "Unterminated String.");
            std::process::exit(65);
        }

        // Find closing "
        self.advance();

        // Strip the surrounding quotes
        let value = &self.source[self.start as usize + 1..self.current as usize - 1];
        let value = value.to_owned();
        self.add_token_with_literal(TokenType::String, Some(LiteralValue::Text(value)));
    }

    // Look one character ahead
    fn peek(&self) -> char {
        if self.is_at_end() { return '\0' }
        return self.source.chars().nth(self.current as usize).unwrap();
    }

    // Select a token type depending on the next token
    fn conditional_select(&mut self, next_token: char, match_type: TokenType, mismatch_type: TokenType) -> TokenType {
        if self.match_next_token(next_token) {
            return match_type;
        } else {
            return mismatch_type;
        }
    }

    // Advance one character ahread
    // Return the character
    fn advance(&mut self) -> char { 
        let c = self.source.chars().nth(self.current as usize).unwrap();
        self.current += 1;
        c
    }

    // Check if the next token matches the expected
    fn match_next_token(&mut self, expected: char) -> bool {

        if self.is_at_end() {
            return false;
        }
        
        if self.source[self.current as usize..].chars().next().unwrap() != expected {
            return false
        }

        self.current += 1;
        true
    }

    // Add a non-literal token
    fn add_token(&mut self, token_type: TokenType) {
        self.add_token_with_literal(token_type, None);
    }

    // Add a token with a literal
    // Ex.
    // x = 42
    // 42 is a literal
    fn add_token_with_literal(&mut self, token_type: TokenType, literal: Option<LiteralValue>) {
       let text = self.source[self.start as usize..self.current as usize].to_string();
       let token = Token::new(token_type, &text, literal, self.line);
       self.tokens.push(token);
    }

    // Check if we have reached the end of source
    fn is_at_end(&self) -> bool {
        self.current >= self.source.len() as i32
    }
}

File: src/main/java/com/craftinginterpreters/lox/Scanner.java
package com.craftinginterpreters.lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Scanner {

    private static final Map<String, TokenType> keywords;
    static {
        keywords = new HashMap<>();
        keywords.put("and", TokenType.AND);
        keywords.put("class", TokenType.CLASS);
        keywords.put("else", TokenType.ELSE);
        keywords.put("false", TokenType.FALSE);
        keywords.put("for", TokenType.FOR);
        keywords.put("fun", TokenType.FUN);
        keywords.put("if", TokenType.IF);
        keywords.put("nil", TokenType.NIL);
        keywords.put("or", TokenType.OR);
        keywords.put("print", TokenType.PRINT);
        keywords.put("return", TokenType.RETURN);
        keywords.put("super", TokenType.SUPER);
        keywords.put("this", TokenType.THIS);
        keywords.put("true", TokenType.TRUE);
        keywords.put("var", TokenType.VAR);
        keywords.put("while", TokenType.WHILE);
    }

    String source;
    List<Token> tokens = new ArrayList<>();
    private int current = 0;

    private int start = 0;

    private int line = 1;

    public Scanner(final String source) {
        this.source = source;
    }

    public List<Token> scanTokens() {
        while (!isEnd()) {
            start = current;
            scanToken();
        }
        tokens.add(new Token(TokenType.EOF, "", null, line));
        return tokens;
    }

    private void scanToken() {
        final char c = advance();

        switch (c) {
            case '(':
                addToken(TokenType.LEFT_PAREN);
                break;
            case ')':
                addToken(TokenType.RIGHT_PAREN);
                break;
            case '{':
                addToken(TokenType.LEFT_BRACE);
                break;
            case '}':
                addToken(TokenType.RIGHT_BRACE);
                break;
            case ',':
                addToken(TokenType.COMMA);
                break;
            case '.':
                addToken(TokenType.DOT);
                break;
            case '-':
                addToken(TokenType.MINUS);
                break;
            case '+':
                addToken(TokenType.PLUS);
                break;
            case ';':
                addToken(TokenType.SEMICOLON);
                break;
            case '*':
                addToken(TokenType.STAR);
                break;
            case '!':
                addToken(matchAdvance('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
                break;
            case '=':
                addToken(matchAdvance('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
                break;
            case '<':
                addToken(matchAdvance('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
                break;
            case '>':
                addToken(matchAdvance('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
                break;
            case '/':
                if (matchAdvance('/')) {
                    while (peek() != '\n' && !isEnd()) {
                        advance();
                    }
                } else {
                    addToken(TokenType.SLASH);
                }
                break;
            case ' ':
            case '\t':
            case '\r':
                break;
            case '\n':
                line++;
                break;
            case '"':
                string();
                break;
            default:
                if (isDigit(c)) {
                    number();
                } else if (isAlpha(c)) {
                    identifier();
                } else {
                    Lox.error(line, "Unexpected character: " + c);
                }
                break;
        }
    }

    private void identifier() {
        while (isAlphanumeric(peek()))
            advance();

        final String text = source.substring(start, current);
        TokenType type = keywords.get(text);

        if (type == null)
            type = TokenType.IDENTIFIER;

        addToken(type);
    }

    private boolean isAlpha(final char c) {
        return (('a' <= c && c <= 'z') ||
                ('A' <= c && c <= 'Z') ||
                (c == '_'));
    }

    private boolean isAlphanumeric(final char c) {
        return isDigit(c) || isAlpha(c);

    }

    private boolean isDigit(final char c) {
        return '0' <= c && c <= '9';
    }

    private void number() {
        while (isDigit(peek()))
            advance();

        if (peek() == '.' && isDigit(peekNext())) {
            advance();

            while (isDigit(peek()))
                advance();
        } else if (peek() == '.') {
            Lox.error(line, "Unexpected character in number");
        }

        addToken(TokenType.NUMBER, Double.parseDouble(source.substring(start, current)));
    }

    private void string() {
        while (peek() != '"' && !isEnd()) {
            if (peek() == '\n')
                line++;
            advance();
        }

        if (isEnd()) {
            Lox.error(line, "Unterminated string.");
            return;
        }

        advance();

        final String value = source.substring(start + 1, current - 1);
        addToken(TokenType.STRING, value);
    }

    private char peek() {
        if (isEnd()) {
            return '\0';
        }

        return source.charAt(current);

    }

    private char peekNext() {
        if (current + 1 >= source.length())
            return '\0';
        return source.charAt(current + 1);
    }

    private boolean matchAdvance(final char value) {
        if (isEnd()) {
            return false;
        }

        if (source.charAt(current) == value) {
            current++;
            return true;
        }

        return false;
    }

    private char advance() {
        return source.charAt(current++);
    }

    private void addToken(final TokenType type) {
        addToken(type, null);
    }

    private void addToken(final TokenType type, final Object literal) {
        final String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

    private boolean isEnd() {
        return current >= source.length();
    }

}
